### PR TITLE
Links

### PR DESCRIPTION
**Pull request by cjm106**
The following has been enabled in this branch:
* The creation of chapter pages based on the markdown files  in the content folder  (the files have the same name as chapter labels to ensure the links work).
* The generation of navbar links based on current chapter labels - next step is to generate links using markdown.

### CODE REVIEW

File: gatsby-node.js [hunk 1]
```diff
@@ -1,25 +1,64 @@
...
+//         }
+//       }
+//     }
+//   `)
+//
+//   if (result.errors) {
+//     reporter.panicOnBuild(`Error while running GraphQL query.`)
+//     return
+//   }
+// }
```
Comment (Gina-Frankel) [inline gatsby-node.js:64]: Delete comments

File: src/templates/chapter.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+export const query = graphql`
+  query($slug: String!) {
+    markdownRemark(fields: { slug: { eq: $slug } }) {
+      html
+      frontmatter {
+        title
+      }
+    }
+  }
+`
```
Comment (Gina-Frankel) [inline src/templates/chapter.js:30]: 👍

File: src/components/navBarItem/navBarItem.tsx [hunk 3]
```diff
@@ -1,10 +1,10 @@
...
-        <p className="navText">{props.chapter}</p>
-      </a>
-    )
-}
\ No newline at end of file
+export const NavBarItem = props => {
+  return (
+    <a className="nav-link" href={props.chapter}>
+      <p className="navText">{props.chapter}</p>
+    </a>
```
Comment (Gina-Frankel) [inline src/components/navBarItem/navBarItem.tsx:8]: `props.chapter could be deconstructed `

File: gatsby-node.js (hunk 1 above)
Comment (cjm106) [reply gatsby-node.js]: removed

File: src/templates/chapter.js (hunk 2 above)
Comment (cjm106) [reply src/templates/chapter.js:30]: It looks to be the preferred option - https://www.gatsbyjs.com/tutorial/part-seven/

File: gatsby-node.js [hunk 4]
```diff
@@ -1,25 +1,39 @@
-exports.createPages = async ({ actions, graphql, reporter }) => {
+const path = require(`path`)
+const { createFilePath } = require(`gatsby-source-filesystem`)
+exports.onCreateNode = ({ node, getNode, actions }) => {
```
Comment (benfurber) [inline gatsby-node.js:3]: For readability let's get a gap between these two lines please.

File: gatsby-node.js [hunk 5]
```diff
@@ -1,25 +1,39 @@
...
+  const { createNodeField } = actions
+  if (node.internal.type === `MarkdownRemark`) {
+    const slug = createFilePath({ node, getNode, basePath: `pages` })
+    createNodeField({
+      node,
+      name: `slug`,
+      value: slug,
+    })
```
Comment (benfurber) [inline gatsby-node.js:11]: Thoughts on doing?:
```
    const basePath: "pages"

    const slug = createFilePath({ node, getNode, basePath })
    const [name, value] = slug

    createNodeField({ node, name, value })
```

File: gatsby-node.js [hunk 6]
```diff
@@ -1,25 +1,39 @@
...
+  }
+}
+exports.createPages = async ({ graphql, actions }) => {
```
Comment (benfurber) [inline gatsby-node.js:14]: Let's get a gap between these lines please.

File: gatsby-node.js [hunk 7]
```diff
@@ -1,25 +1,39 @@
...
-  if (result.errors) {
-    reporter.panicOnBuild(`Error while running GraphQL query.`)
-    return
-  }
+  result.data.allMarkdownRemark.edges.forEach(({ node }) => {
+    createPage({
+      path: node.fields.slug,
+      component: path.resolve(`./src/templates/chapter.js`),
+      context: {
+        slug: node.fields.slug,
```
Comment (benfurber) [inline gatsby-node.js:35]: You're calling `node.fields.slug` twice. How can you dry this up?

File: gatsby-node.js [hunk 8]
```diff
@@ -1,25 +1,39 @@
...
   `)
 
-  if (result.errors) {
-    reporter.panicOnBuild(`Error while running GraphQL query.`)
-    return
-  }
+  result.data.allMarkdownRemark.edges.forEach(({ node }) => {
+    createPage({
+      path: node.fields.slug,
+      component: path.resolve(`./src/templates/chapter.js`),
```
Comment (benfurber) [inline gatsby-node.js:33]: With `path.resolve(`./src/templates/chapter.js`)` on this line you're doing more than I'd like on the line. Any thoughts on a different way of doing this?

File: src/components/navBarItem/navBarItem.tsx [hunk 9]
```diff
@@ -1,10 +1,11 @@
...
 
-export const NavBarItem = (props) => {
-    return (
-        <a className="nav-link" href="/chapter1">
-        <p className="navText">{props.chapter}</p>
-      </a>
-    )
-}
\ No newline at end of file
+export const NavBarItem = props => {
```
Comment (benfurber) [inline src/components/navBarItem/navBarItem.tsx:4]: conscious decision to not add a type def here?

File: src/templates/chapter.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
+import React from "react"
+import { graphql } from "gatsby"
+import { Layout } from "../components"
+import { DoublePage } from "../components/index"
+
+export default function NewChapter({ data }) {
```
Comment (benfurber) [inline src/templates/chapter.js:6]: Conscious decision to not add a type def here?

File: src/templates/chapter.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
+import React from "react"
+import { graphql } from "gatsby"
+import { Layout } from "../components"
```
Comment (benfurber) [inline src/templates/chapter.js:3]: Line break for readability please. I like to seperate out external imports, from project imports.

File: src/templates/chapter.js [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+export default function NewChapter({ data }) {
+  const post = data.markdownRemark
+  return (
+    <div>
+      <Layout>
+        <DoublePage />
+        <div>
+          <h1>{post.frontmatter.title}</h1>
+          <div dangerouslySetInnerHTML={{ __html: post.html }} />
```
Comment (benfurber) [inline src/templates/chapter.js:14]: Have you researched other ways than to do a dangerous html call here?

File: src/templates/chapter.js [hunk 13]
```diff
@@ -0,0 +1,30 @@
+import React from "react"
+import { graphql } from "gatsby"
+import { Layout } from "../components"
+import { DoublePage } from "../components/index"
+
+export default function NewChapter({ data }) {
+  const post = data.markdownRemark
```
Comment (benfurber) [inline src/templates/chapter.js:7]: I'd prefer:
```
   const { post } = data.markdownRemark
```

Though you can actually go one step further (then change the implimentation further on):
```
   const { frontmatter, html } = data.markdownRemark.post
   const { title } = frontmatter
```

File: src/templates/chapter.js (hunk 2 above)
Comment (benfurber) [inline src/templates/chapter.js:30]: We're starting to collect a few graphql queries now, we might want to centralise them soon. Not neccessarily now.

**Review by benfurber — CHANGES_REQUESTED** [review]

File: gatsby-node.js (hunk 4 above)
Comment (cjm106) [reply gatsby-node.js:3]: line break added

File: gatsby-node.js (hunk 6 above)
Comment (cjm106) [reply gatsby-node.js:14]: done

File: gatsby-node.js (hunk 5 above)
Comment (cjm106) [reply gatsby-node.js:13]: I'm unable to get your suggested chanegs to work - even a partial i.e.    
const [name, value] = slug
    createNodeField({
      node,
      name,
      value,
    })

File: gatsby-node.js (hunk 7 above)
Comment (cjm106) [reply gatsby-node.js]: result.data.allMarkdownRemark.edges.forEach(({ node }) => {
    const field = node.fields.slug
    createPage({
      path: field,
      component: path.resolve(`./src/templates/chapter.js`),
      context: {
        slug: field,
      },
    })
  })

File: gatsby-node.js (hunk 8 above)
Comment (cjm106) [reply gatsby-node.js:37]: The gatsby-node file is based on the example provided here https://www.gatsbyjs.com/tutorial/part-seven/ - I will need to research this further.  

Are you suggesting something like: 

import MyComponent from 'components/MyComponent
...
component: MyComponent

File: src/templates/chapter.js (hunk 11 above)
Comment (cjm106) [reply src/templates/chapter.js:3]: added

File: src/components/navBarItem/navBarItem.tsx (hunk 9 above)
Comment (cjm106) [reply src/components/navBarItem/navBarItem.tsx:4]: Yes, I didnt want to do something that might duplicate/conflict with @mattfreeman-london's navBarTest branch

File: src/templates/chapter.js (hunk 12 above)
Comment (cjm106) [reply src/templates/chapter.js]: Not at this stage, current implementation is based on gatsbyjs.com documentation/tutorial. Would you prefer something similar to this?
var test = React.createClass({
  componentDidUpdate: function(prevProp, prevState){
    this.refs.test.innerHTML = "Hello";
  },
  render: function(){
    return (
      <div contentEditable='true' ref='test'></div>
    );
  }
});

File: src/templates/chapter.js (hunk 13 above)
Comment (cjm106) [reply src/templates/chapter.js]: Both options appear to break things at this stage, producing :

'TypeError: Cannot read property 'frontmatter' of undefined'

File: src/templates/chapter.js (hunk 2 above)
Comment (cjm106) [reply src/templates/chapter.js]: Will add it to the project board.

File: src/templates/chapter.js (hunk 10 above)
Comment (cjm106) [reply src/templates/chapter.js]: Yes. I've converted the file to .tsx  and have defined data as follows
interface Props {
  data: object
}
Is this the correct way to define the typ def?

File: src/templates/chapter.js (hunk 13 above)
Comment (Gina-Frankel) [reply src/templates/chapter.js]: I think this may be because there is no property of **post** in data

Comment (Gina-Frankel) [reply src/templates/chapter.js]: image

Comment (Gina-Frankel) [reply src/templates/chapter.js]: solution: 
` const { frontmatter, html } = data.markdownRemark
  const { title } = frontmatter`

File: gatsby-node.js (hunk 5 above)
Comment (benfurber) [reply gatsby-node.js:13]: Ugh. Sorry, I miss typed. 
```
 const [name, value] = [slug]
```
Should do it. I'm honestly on the fence about it anyway, the syntax is a little weird but I like the consistency then for `createNodeField`.

File: gatsby-node.js (hunk 8 above)
Comment (benfurber) [reply gatsby-node.js]: No, I was thinking:
```js
const component = path.resolve(`./src/templates/chapter.js`)
```

File: src/templates/chapter.js (hunk 10 above)
Comment (benfurber) [reply src/templates/chapter.js]: Yes, though ideally you'd also define what the object is.

File: src/templates/chapter.js (hunk 12 above)
Comment (benfurber) [reply src/templates/chapter.js]: It's the `dangerouslySetInnerHTML` bit I don't really like - though the data isn't coming from a data source the author doesn't have complete control of.  Basically the reason it's dangerous is that it'll allow ANY html tags so it's vunerable to security attacks.

**Pull request merged**